kotlinx-coroutines-core/common/src/sync/Mutex.kt

```diff
@@ -361,24 +358,28 @@ internal class MutexImpl(locked: Boolean) : Mutex, SelectClause2<Any?, Mutex> {
     }
 
     private class LockCont(
+        mutex: MutexImpl,
         owner: Any?,
         @JvmField val cont: CancellableContinuation<Unit>
-    ) : LockWaiter(owner) {
-        override fun tryResumeLockWaiter() = cont.tryResume(Unit)
+    ) : LockWaiter(mutex, owner) {
+        override fun tryResumeLockWaiter() = cont.tryResume(Unit, idempotent = null) {
```

Comment:
You can avoid allocation here by making `LockCont` implement `(cause: Throwable) -> Unit` and passing `this` to the second parameter

Reply:
Cannot do it. It will not work on JS.